### PR TITLE
feat(log_to_metric transform): support tags expansion similar to `labels expansion` for loki sink

### LINKED ISSUE #14744: logs_to_metric transform should allow for deriving tags field from a field (without specifically listing all the possible keys)
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

It appears that the `log_to_metric` transform has no way to support deriving an arbitrary set of tags from a log event ...?

A configuration like this will fail because the `tags` configuration property is required to be a map ...

```
    metrics:
      - type: "histogram"
        field: "FIELD"
        name: "{{NAME}}"
        namespace: "{{NAMESPACE}}"
        tags: "{{TAGS}}"
```

### Attempted Solutions

- I tried with and without the template expansion
- I looked  into whether I could use remap to add the tags back onto the metric after it had been converted from log into a metric -- but I'm not clear on whether its possible to pass information through the pipeline in such a way to allow this ...

I've ended up working around this issue for now with a lua transform  ... -- at the moment I only care about supporting `gauge` metrics and so I've only implemented the transform for the gauge type (I suspect there is likely more sophisticated logic needed for some of the other metrics ...)

It would _really_ nice if I could instead just point the `logs_to_metric` transform at a field containing all the tags I want defined on the output metric ...  To prevent high cardinality metric issues -- I'd use the `tag_cardinality_limit` transform ...

I'm working around for now with a lua transform.  I'm filing this feature request to document a use case where lua transform seems required (as suggested in the [lua transform docs](https://vector.dev/docs/reference/configuration/transforms/lua/))
 
```
transforms:
  parsing:
    type: "remap"
    inputs:
      - snowflake_s3
    source: |-
      . = parse_json!(string!(.message))

  route_logs_by_type:
    type: route
    inputs:
      - parsing
    route:
      counter: .TYPE == "counter"
      histogram: .TYPE == "histogram"
      gauge: .TYPE == "gauge"
      set: .TYPE == "set"
      summary: .TYPE == "summary"
      log: .TYPE == "log"

  remap_counter_log_to_metric:
    type: log_to_metric
    inputs:
      - route_logs_by_type.counter
    metrics:
      - type: "counter"
        field: "FIELD"
        name: "{{NAME}}"
        namespace: "{{NAMESPACE}}"

  remap_histogram_log_to_metric:
    type: log_to_metric
    inputs:
      - route_logs_by_type.histogram
    metrics:
      - type: "histogram"
        field: "FIELD"
        name: "{{NAME}}"
        namespace: "{{NAMESPACE}}"

  # {"name":"storage.table.retained_bytes.avg","namespace":"snowflake","tags":{"env":"dev","schema":"STAGING","service":"snowflake","table":"SOME_LOG"},"timestamp":"2022-10-05T21:09:16Z","kind":"absolute","gauge":{"value":0.0}}
  remap_gauge_log_to_metric:
    type: lua
    version: "2"
    inputs:
      - route_logs_by_type.gauge
    hooks:
      process: |-
        function (event, emit)
          event.metric = {
            name = event.log.NAME,
            namespace = event.log.NAMESPACE,
            kind = "absolute",
            timestamp = os.date("!*t"),
            tags = event.log.TAGS,
            gauge = {
              value = event.log.FIELD
            }
          }
          event.log = nil
          emit(event)
        end

  # {"name":"storage.table.retained_bytes.avg","namespace":"snowflake","timestamp":"2022-10-05T21:04:04.776646Z","kind":"absolute","gauge":{"value":0.0}}
  # remap_gauge_log_to_metric:
  #   type: log_to_metric
  #   inputs:
  #     - route_logs_by_type.gauge
  #   metrics:
  #     - type: "gauge"
  #       field: "FIELD"
  #       name: "{{NAME}}"
  #       namespace: "{{NAMESPACE}}"

  remap_set_log_to_metric:
    type: log_to_metric
    inputs:
      - route_logs_by_type.set
    metrics:
      - type: "gauge"
        field: "FIELD"
        name: "{{NAME}}"
        namespace: "{{NAMESPACE}}"

  remap_summary_log_to_metric:
    type: log_to_metric
    inputs:
      - route_logs_by_type.summary
    metrics:
      - type: "summary"
        field: "FIELD"
        name: "{{NAME}}"
        namespace: "{{NAMESPACE}}"

  remap_log_field_to_message:
    type: remap
    inputs:
      - route_logs_by_type.log
    source: |-
      .message = .FIELD
      del(.FIELD)
      .ddsource = .NAMESPACE
      del(.NAMESPACE)
      .service = .TAGS.service
      .ddtags = .TAGS
      del(.TAGS)
      del(.ddtags.service)
      del(.TYPE)
      .timestamp = now()

sinks:
  datadog_metrics:
    type: datadog_metrics
    inputs:
      - remap_counter_log_to_metric
      - remap_histogram_log_to_metric
      - remap_gauge_log_to_metric
      - remap_set_log_to_metric
      - remap_summary_log_to_metric
    default_api_key: "${DD_API_KEY:?err}"

  datadog_logs:
    type: datadog_logs
    inputs:
      - remap_log_field_to_message
      - route_logs_by_type._unmatched
    default_api_key: "${DD_API_KEY:?err}"

  console_output:
    type: console
    inputs:
      - remap*
    encoding:
      codec: "text"

tests:
  - name: "parsing -> parsing"
    inputs:
      - type: raw
        insert_at: parsing
        value: |-
          {"FIELD":0,"NAME":"storage.table.retained_bytes.avg","NAMESPACE":"snowflake","TAGS":{"env":"dev","schema":"STAGING","service":"snowflake","table":"STORAGE_LOG"},"TYPE":"gauge"}

    outputs:
      - extract_from: parsing
        conditions:
          - type: vrl
            source: |-
              assert!(exists(.TYPE))

  - name: "parsing -> route_logs_by_type.gauge"
    inputs:
      - type: raw
        insert_at: parsing
        value: |-
          {"FIELD":0,"NAME":"storage.table.retained_bytes.avg","NAMESPACE":"snowflake","TAGS":{"env":"dev","schema":"STAGING","service":"snowflake","table":"STORAGE_LOG"},"TYPE":"gauge"}

    outputs:
      - extract_from: route_logs_by_type.gauge
        conditions:
          - type: vrl
            source: |-
              assert!(exists(.NAMESPACE))

  - name: "parsing -> remap_gauge_log_to_metric"
    inputs:
      - type: raw
        insert_at: parsing
        value: |-
          {"FIELD":0,"NAME":"storage.table.retained_bytes.avg","NAMESPACE":"snowflake","TAGS":{"env":"dev","schema":"STAGING","service":"snowflake","table":"STORAGE_LOG"},"TYPE":"gauge"}

    outputs:
      # {"name":"storage.table.retained_bytes.avg","namespace":"snowflake","tags":{"env":"dev","schema":"STAGING","service":"snowflake","table":"STORAGE_LOG"},"timestamp":"2022-10-05T21:09:16Z","kind":"absolute","gauge":{"value":0.0}}
      - extract_from: remap_gauge_log_to_metric
        conditions:
          - type: vrl
            source: |-
              assert!(exists(.name))
              assert!(exists(.namespace))
              assert!(exists(.tags))
              assert!(exists(.timestamp))
              assert!(exists(.kind))

  - name: "parsing -> remap_log_field_to_message"
    inputs:
      - type: raw
        insert_at: parsing
        value: |-
          {"FIELD":"test logging","NAMESPACE":"snowflake","TAGS":{"env":"dev","service":"snowflake"},"TYPE":"log"}

    outputs:
      # {"ddsource":"snowflake","ddtags":{"env":"dev"},"message":"test logging","service":"snowflake","timestamp":"2022-10-05T21:48:30.823011Z"}
      - extract_from: remap_log_field_to_message
        conditions:
          - type: vrl
            source: |-
              assert!(exists(.ddsource))
              assert!(exists(.ddtags))
              assert!(exists(.message))
              assert!(exists(.service))
              assert!(exists(.timestamp))
```

### Proposal

Expand `log_to_metric` with a new parameter that allows deriving all tags for the metric from a particular field in the log.  The field in the log will be expected and required to be a map from string -> string.

### References

- https://github.com/vectordotdev/vector/issues/14645#issuecomment-1269027111

### Version

vector 0.24.1 (x86_64-apple-darwin 8935681 2022-09-12)

### PR DESCRIPTION
**Pull request by titaneric**
## Summary

This PR aims to support `tags expansion` for `log_to_metric transform` similar [Loki's sink's labels expansion](https://vector.dev/docs/reference/configuration/sinks/loki/#label-expansion). As a new vector contributor, I copy the code from Loki sink and add unit test cases for these changes. If the further code refactor or docs are needed, I would immediately apply the changes. I would like to hear there's a more better implementation to this feature, and I am welcome to hear any comments.

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?

make test

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [x] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- [x] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

Close #14744.

Partially solve #15332 ?

### CODE REVIEW

**Comment by pront** [comment]
Hi @titaneric, thank you for this contribution. I see this is marked as draft, feel free to give me a shout when you want a review.

**Comment by titaneric** [comment]
There is a testcase named `variants::disk_v2::tests::invariants::reader_deletes_data_file_around_record_id_wraparound` did not pass, and it seems that it's not related to my change.

I have added new testcase borrowed from loki's sink. @pront  Please take your time to review it.

File: src/transforms/log_to_metric.rs [hunk 1]
```diff
@@ -271,55 +275,150 @@ fn render_template(template: &Template, event: &Event) -> Result<String, Transfo
...
+                                &mut dynamic_tags,
+                            )?;
                         }
                     }
                 }
             }
+            for (k, v) in static_tags {
+                if let Some(discarded_v) = dynamic_tags.insert(k.clone(), v.clone()) {
+                    warn!(
+                        "Static static overrides dynamic tag. \
```
Comment (pront) [inline src/transforms/log_to_metric.rs:316]:
```suggestion
                        "Static tags overrides dynamic tags. \
```

File: changelog.d/21939_log_to_metric_tags_expansion.feature.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+The `log_to_metric` transformer now support tags expansion similar to Loki sink's labels expansion.
```
Comment (pront) [inline changelog.d/21939_log_to_metric_tags_expansion.feature.md:1]:
```suggestion
The `log_to_metric` transformer tag key are now template-able which enables tags expansion.
```

File: src/transforms/log_to_metric.rs [hunk 3]
```diff
@@ -271,55 +275,150 @@ fn render_template(template: &Template, event: &Event) -> Result<String, Transfo
...
 }
 
+static RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^0-9A-Za-z_]").unwrap());
+fn slugify_text(input: String) -> String {
+    let result = RE.replace_all(&input, "_");
+    result.to_lowercase()
+}
+
+// borrow from src/sinks/loki/sink.rs
+fn tag_expansion(
```
Comment (pront) [inline src/transforms/log_to_metric.rs:376]: I didn't look deeply yet but it seems like we can benefit from introducing a util, maybe under `src/common` to avoid code duplication.

**Review by pront — COMMENTED** [review]
Thanks @titaneric, this change makes sense to me 👍 

Left a few comments. Also, please run `make generate-component-docs`

File: src/transforms/log_to_metric.rs (hunk 3 above)
Comment (titaneric) [reply src/transforms/log_to_metric.rs:376]: It's a good idea! I could try it.

File: src/sinks/loki/sink.rs [hunk 4]
```diff
@@ -284,46 +249,12 @@ impl EventEncoder {
...
-                            "Encountered duplicated dynamic structured_metadata. \
-                        key: {}, value: {}, discarded value: {}",
-                            key, val, prev
-                        );
-                    };
-                }
-            } else {
-                static_structured_metadata.insert(key_s, value_s);
-            }
+            let _ = pair_expansion(
```
Comment (titaneric) [inline src/sinks/loki/sink.rs:252]: I don't know how to handle the result if `pair_expansion` has error since we have already printed the warning message in `pair_expansion` function. I just simply drop the result here. Please give any advices if there is a better solution.

**Comment by titaneric** [comment]
I have moved the core expansion function to `common` mod, and adopt them in both loki sink and log_to_metrics transforms. Please review them if you are available!

File: website/cue/reference/components/transforms/base/log_to_metric.cue [hunk 5]
```diff
@@ -85,8 +85,12 @@ base: components: transforms: log_to_metric: configuration: {
 				type: string: syntax: "template"
 			}
 			tags: {
-				description: "Tags to apply to the metric."
-				required:    false
+				description: """
+					Tags to apply to the metric.
+
+					Both keys and values are templateable, which enables you to attach dynamic tags to events.
```
Comment (drichards-87) [inline website/cue/reference/components/transforms/base/log_to_metric.cue:91]:
```suggestion
					Both keys and values can be templated, allowing you to attach dynamic tags to events.
```

**Review by drichards-87 — APPROVED** [review]
Left a very minor suggestion from Docs and approved the PR.

File: website/cue/reference/components/transforms/base/log_to_metric.cue (hunk 5 above)
Comment (titaneric) [reply website/cue/reference/components/transforms/base/log_to_metric.cue]: Hi, I update the docs in source code and re-generate the docs. Thank you for your advice!

**Comment by titaneric** [comment]
@pront Would you please take a look at this PR when you are available?

**Comment by pront** [comment]
> @pront Would you please take a look at this PR when you are available?

Yes, it's on my radar.

File: src/common/expansion.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
+use regex::Regex;
+use std::{collections::HashMap, sync::LazyLock};
+
+use crate::event::Value;
+
+static RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^0-9A-Za-z_]").unwrap());
+fn slugify_text(input: String) -> String {
```
Comment (pront) [inline src/common/expansion.rs:7]:
```suggestion
fn slugify_text(input: &str) -> String {
```

File: src/common/expansion.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    let result = RE.replace_all(&input, "_");
+    result.to_lowercase()
+}
+
+pub(crate) fn pair_expansion(
+    key_s: String,
+    value_s: String,
```
Comment (pront) [inline src/common/expansion.rs:14]:
```suggestion
    key_s: &str,
    value_s: &str,
```

This will require some usage updates.

File: src/sinks/loki/sink.rs (hunk 4 above)
Comment (pront) [reply src/sinks/loki/sink.rs:252]: It looks like the behavior is equivalent so dropping the result here is fine since there's nothing else to do.

File: src/transforms/log_to_metric.rs [hunk 8]
```diff
@@ -271,52 +276,96 @@ fn render_template(template: &Template, event: &Event) -> Result<String, Transfo
...
-                    error,
+                    error: value_error,
                     drop_event: false,
-                    field: Some(name),
+                    field: Some(template.get_ref()),
                 });
                 return Ok(());
             }
-            Err(other) => return Err(other),
+            Err(err) => return Err(err),
```
Comment (pront) [inline src/transforms/log_to_metric.rs:366]: Tiny nit: 
```suggestion
            Err(other) => return Err(other),
```

File: src/transforms/log_to_metric.rs [hunk 9]
```diff
@@ -271,52 +276,96 @@ fn render_template(template: &Template, event: &Event) -> Result<String, Transfo
...
         }
     })
 }
 
 fn render_tag_into(
     event: &Event,
-    name: &str,
-    template: &Option<Template>,
+    key_template: &Template,
+    value_template: &Option<Template>,
```
Comment (pront) [inline src/transforms/log_to_metric.rs:331]: I understand this is existing code but we are trying to switch to the following pattern:
```suggestion
    value_template: Option<&Template>,
```

**Review by pront — COMMENTED** [review]
Thank you @titaneric, this is looking good. Left some suggestions for minor improvements.

File: src/sinks/loki/sink.rs [hunk 10]
```diff
@@ -179,42 +179,7 @@ impl EventEncoder {
...
-                            "Encountered duplicated dynamic label. \
-                                key: {}, value: {}, discarded value: {}",
-                            key, val, prev
-                        );
-                    };
-                }
-            } else {
-                static_labels.insert(key_s, value_s);
-            }
+            let _ = pair_expansion(key_s, value_s, &mut static_labels, &mut dynamic_labels);
```
Comment (pront) [inline src/sinks/loki/sink.rs:182]: Hm, should we check if this returned an error and if so, `continue;`? Let's be extra careful to not change the `loki` sink behavior.

Comment (titaneric) [reply src/sinks/loki/sink.rs]: I guess we could leave it unchanged? It's basically do the same thing as expanding `structured_metadata`.

Comment (pront) [reply src/sinks/loki/sink.rs]: I am referring to this existing code: https://github.com/vectordotdev/vector/blob/master/src/sinks/loki/sink.rs#L195

It's hard to follow without checking it out locally (too many nested blocks in this code).
As long as we figure this out this PR is good to go.

**Review by pront — APPROVED** [review]

File: src/common/expansion.rs [hunk 11]
```diff
@@ -4,21 +4,24 @@ use std::{collections::HashMap, sync::LazyLock};
...
 static RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^0-9A-Za-z_]").unwrap());
-fn slugify_text(input: String) -> String {
+fn slugify_text(input: &str) -> String {
     let result = RE.replace_all(&input, "_");
     result.to_lowercase()
 }
 
+/// Expands the given possibly template-able `key_s` and `value_s`, and return the expanded owned pairs
+/// it would also insert the pairs into either `static_pairs` or `dynamic_pairs` depending on the template-ability of `key_s`.
+/// Refer to loki sink and log_to_metric transform for further information.
```
[the remark this answers is not in the document]
Comment (titaneric) [reply src/common/expansion.rs:14]: Sorry, do you mean line 14 or the whole function comment?

**Comment by titaneric** [comment]
It appears that test suite does not pass, let me fix it and double-check the equivalence of changed code.

File: src/sinks/loki/sink.rs (hunk 10 above)
Comment (titaneric) [reply src/sinks/loki/sink.rs]: ok, let me double-check it

Comment (titaneric) [reply src/sinks/loki/sink.rs]: I think they are equivalent. The original loki sink continues the next iteration while having the serde error in the for loop. On the other hand, my implementation return the error from the `pair_expansion` function in the for loop. As a result, they do not continue to make further progress while having the serde error.

Honestly, I think loki sink needs some comments to explain, or handle the error properly instead of dropping it.

**Comment by titaneric** [comment]
Oh, I just find that I forget to update the document, and I will commit it later.

**Comment by titaneric** [comment]
I think `log_to_metric` transform deserve another PR to update the documents.

**Comment by titaneric** [comment]
I have updated the example to adopt tags expansion feature

Not sure whether it's a known issue or not, I notice that `metric` option could not be properly rendered on the documents, and it appears that it also affect the [static_metrics](https://vector.dev/docs/reference/configuration/sources/static_metrics/) docs. It seems that the page builder could not support **array of object** in cue, and I would open another issue to report it. 

preview page:

image

cue:

image

**Comment by pront** [comment]
> It appears that test suite does not pass, let me fix it and double-check the equivalence of changed code.

I think your latest changes fixed this.

**Comment by titaneric** [comment]
Hi @pront , is it possible this PR would be merged in the upcoming days?

File: src/sinks/loki/sink.rs (hunk 10 above)
Comment (pront) [reply src/sinks/loki/sink.rs]: > Honestly, I think loki sink needs some comments to explain, or handle the error properly instead of dropping it

Agreed!

File: src/common/expansion.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (pront) [reply src/common/expansion.rs]: Yes, just line 14.

**Comment by titaneric** [comment]
@pront , thanks for merging this PR, and I just find another related issue at https://github.com/vectordotdev/vector/issues/7565 which I believed it could be closed!
